action/protocol/rolldpos/epoch_test.go

```diff
@@ -1,36 +1,187 @@
 package rolldpos
 
 import (
+	"context"
 	"testing"
 
 	"github.com/stretchr/testify/require"
 )
 
-func TestGetEpochNum(t *testing.T) {
+
+func TestEnableDardanellesSubEpoch(t *testing.T) {
 	require := require.New(t)
+
+	height := 0
+	numSubEpochs := 1
+
+	options := EnableDardanellesSubEpoch(uint64(height), uint64(numSubEpochs))
 	p := NewProtocol(23, 4, 3)
-	epochHeights := []uint64{0, 1, 12, 25, 38, 53, 59, 80, 90, 93, 120}
-	expectedNums := []uint64{0, 1, 1, 3, 4, 5, 5, 7, 8, 8, 10}
-	for i, epochHeight := range epochHeights {
-		num := p.GetEpochNum(epochHeight)
-		require.Equal(expectedNums[i], num)
+	require.Nil(options(p))
+	require.NotNil(options)
+}
+
+func TestNewProtocol(t *testing.T) {
+	require := require.New(t)
+	numCandidateDelegates := uint64(23)
+	numDelegates := uint64(24)
+	numSubEpochs := uint64(3)
+	height := -1
+	options := EnableDardanellesSubEpoch(uint64(height), numSubEpochs)
+	p := NewProtocol(numCandidateDelegates, numDelegates, numSubEpochs, options)
+	require.NotNil(p)
+
+}
+
+func TestProtocol_Handle(t *testing.T) {
+	require := require.New(t)
+	p := NewProtocol(23, 4, 3)
+	ctx := context.Background()
+	require.Nil(p.Handle(ctx, nil, nil))
+
+}
+
+func TestProtocol_Validate(t *testing.T) {
+	require := require.New(t)
+	p := NewProtocol(23, 4, 3)
+	ctx := context.Background()
+	require.Nil(p.Validate(ctx, nil))
+}
+
+func TestProtocol_NumCandidateDelegates(t *testing.T) {
+	require := require.New(t)
+	p := NewProtocol(23, 4, 3)
+	require.Equal(uint64(23), p.NumCandidateDelegates())
+}
+
+func TestProtocol_NumDelegates(t *testing.T) {
+	require := require.New(t)
+	p := NewProtocol(23, 4, 3)
+	require.Equal(uint64(4), p.NumDelegates())
+}
+
+func TestProtocol_ReadState(t *testing.T) {
+	require := require.New(t)
+	p := NewProtocol(23, 4, 3)
+
+	ctx := context.Background()
+	methods := [8]string{
+		"NumCandidateDelegates",
+		"NumDelegates",
+		"NumSubEpochs",
+		"EpochNumber",
+		"EpochHeight",
+		"EpochLastHeight",
+		"SubEpochNumber",
+		"trick",
+	}
+	for _, method := range methods {
+		_, err:=p.ReadState(ctx, nil, []byte(method), []byte("trick_arg_1"), []byte("trick_arg_2"))
+		require.Error(err)
 	}
+	for _, method := range methods {
+		_, err:=p.ReadState(ctx, nil, []byte(method), []byte("good_arg_1"))
```

Comment:
if possible, let's validate the return value as well. Or we can leave a TODO here, and leave it to next PR.